storagenode/piecestore/endpoint.go

```diff
@@ -181,6 +182,28 @@ func (endpoint *Endpoint) DeletePieces(
 	}, nil
 }
 
+func (endpoint *Endpoint) flagUnusualSlowUpload(averageUploadRate float64, connectionCount int64) bool {
+	// congestionThreshold defines the proportion of MaxConcurrentRequests
+	// that must be reached before the storage node will begin dropping slow
+	// upload connections to preserve its own bandwidth.
+	const congestionThreshold = 0.8
+
+	if averageUploadRate < float64(endpoint.config.MinUploadSpeed) {
+		if endpoint.config.MaxConcurrentRequests != 0 {
+			// if 80% of the connection pool is occupied and average upload falls below limit
+			// then the connection is flagged
+			requestCongestionThreshold := int64(float64(endpoint.config.MaxConcurrentRequests) * congestionThreshold)
+
+			if (connectionCount > requestCongestionThreshold) && averageUploadRate < float64(endpoint.config.MinUploadSpeed) {
```

Comment:
Note, here we need to use latest connection count, because this may change during the upload.

Reply:
Done in `isSlowUpload` function